dali/pipeline/executor/executor2/exec2.cc

```diff
@@ -0,0 +1,459 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <functional>
+#include <map>
+#include <queue>
+#include <unordered_map>
+#include <utility>
+#include "dali/core/cuda_stream_pool.h"
+#include "dali/pipeline/executor/executor2/exec2.h"
+#include "dali/pipeline/executor/executor2/exec_graph.h"
+#include "dali/pipeline/executor/executor2/stream_assignment.h"
+
+namespace dali {
+namespace exec2 {
+
+namespace {
+
+void LimitBackendConcurrency(ExecGraph &graph, OpType backend, int max_concurrency = 1) {
+  auto sem = std::make_shared<tasking::Semaphore>(max_concurrency);
+  for (auto &n : graph.Nodes()) {
+    if (n.backend == backend)
+        n.concurrency = sem;
+  }
+  graph.Invalidate();
+}
+
+void ApplyConcurrencyLimit(ExecGraph &graph, OperatorConcurrency concurrency) {
+  switch (concurrency) {
+    case OperatorConcurrency::Full:
+      // TODO(michalz): Fix ThreadPool.
+      LimitBackendConcurrency(graph, OpType::CPU);
+      break;  // other operators have no restrictions
+    case OperatorConcurrency::Backend:
+      LimitBackendConcurrency(graph, OpType::CPU);
+      LimitBackendConcurrency(graph, OpType::GPU);
+      LimitBackendConcurrency(graph, OpType::MIXED);
+      break;
+    case OperatorConcurrency::None:
+      {
+        auto sem = std::make_shared<tasking::Semaphore>(1);
+        for (auto &n : graph.Nodes())
+            n.concurrency = sem;
+      }
+      break;
+    default:
+      assert(!"Unexpected concurrency policy value.");
+      break;
+  }
+}
+
+}  // namespace
+
+class Executor2::Impl {
+ public:
+  explicit Impl(const Config &config) : config_(config) {
+  }
+
+  ~Impl() {
+    dtor_guard_.emplace(config_.device.value_or(CPU_ONLY_DEVICE_ID));
+    Shutdown();
+  }
+
+  enum class State {
+    New = 0,
```

Comment:
Do you expect to have more states in the future (if not ignore the comment)? If so, I would ask myself if any transitions could be in invalid, e.g. (Building->Running)? I guess I would be happy to see a state transition method, that would throw an exception if advancing between specific states is not correct. This would give an opportunity to find such bugs.